protobuf/lib/src/protobuf/pb_list.dart

```diff
@@ -6,310 +6,217 @@ part of protobuf;
-
-  @override
-  Iterable<E> take(int count) => _wrappedList.take(count);
-
-  @override
-  Iterable<E> takeWhile(bool Function(E value) test) =>
-      _wrappedList.takeWhile(test);
-
-  @override
-  Iterable<E> skip(int count) => _wrappedList.skip(count);
-
-  @override
-  Iterable<E> skipWhile(bool Function(E value) test) =>
-      _wrappedList.skipWhile(test);
-
-  @override
-  E get first => _wrappedList.first;
-
-  @override
-  E get last => _wrappedList.last;
-
-  @override
-  E get single => _wrappedList.single;
-
-  @override
-  E firstWhere(bool Function(E element) test, {E Function()? orElse}) =>
-      _wrappedList.firstWhere(test, orElse: orElse);
-
-  @override
-  E lastWhere(bool Function(E element) test, {E Function()? orElse}) =>
-      _wrappedList.lastWhere(test, orElse: orElse);
-
-  @override
-  E elementAt(int index) => _wrappedList.elementAt(index);
-
-  @override
-  String toString() => _wrappedList.toString();
-
   @override
   E operator [](int index) => _wrappedList[index];
 
   @override
-  int get length => _wrappedList.length;
-
-  // TODO(jakobr): E instead of Object once dart-lang/sdk#31311 is fixed.
-  @override
-  int indexOf(Object? element, [int start = 0]) =>
-      _wrappedList.indexOf(element as E, start);
-
-  // TODO(jakobr): E instead of Object once dart-lang/sdk#31311 is fixed.
-  @override
-  int lastIndexOf(Object? element, [int? start]) =>
-      _wrappedList.lastIndexOf(element as E, start);
+  void operator []=(int index, E value) {
+    if (_isReadOnly) {
+      throw UnsupportedError('Setting field of a read-only list');
+    }
+    _check(value);
+    _wrappedList[index] = value;
+  }
 
   @override
-  List<E> sublist(int start, [int? end]) => _wrappedList.sublist(start, end);
+  bool operator ==(other) => other is PbList && _areListsEqual(other, this);
 
   @override
-  Iterable<E> getRange(int start, int end) => _wrappedList.getRange(start, end);
+  int get hashCode => _HashUtils._hashObjects(_wrappedList);
 
-  @override
-  Map<int, E> asMap() => _wrappedList.asMap();
+  void freeze() {
+    if (_isReadOnly) {
+      return;
+    }
 
-  @override
-  void operator []=(int index, E value) {
-    check(value);
-    _wrappedList[index] = value;
-  }
+    _isReadOnly = true;
 
-  /// Unsupported -- violated non-null constraint imposed by protobufs.
-  ///
-  /// Changes the length of the list. If [newLength] is greater than the current
-  /// [length], entries are initialized to [:null:]. Throws an
-  /// [UnsupportedError] if the list is not extendable.
-  @override
-  set length(int newLength) {
-    if (newLength > length) {
-      throw UnsupportedError('Extending protobuf lists is not supported');
+    // Per spec `repeated map<..>` and `repeated repeated ..` are not allowed
+    // so we only check for messages
+    if (_isGroupOrMessage(_elementType)) {
+      for (var elem in _wrappedList) {
+        (elem as GeneratedMessage).freeze();
+      }
     }
```

Comment:
Another implementation of this would be:

```suggestion
    if (_wrappedList.isNotEmpty && _wrappedList[0] is GeneratedMessage) {
      for (var elem in _wrappedList) {
        (elem as GeneratedMessage).freeze();
      }
    }
```

Now we don't need to store the element type anymore.